src/internal/connector/onedrive/url_cache.go

```diff
@@ -0,0 +1,275 @@
+			errors:          fault.New(false),
+		},
+		nil
+}
+
+// validateCacheParams validates the parameters passed to newURLCache
+func validateCacheParams(
+	driveID string,
+	refreshInterval time.Duration,
+	svc graph.Servicer,
+	itemPagerFunc func(
+		servicer graph.Servicer,
+		driveID, link string,
+	) itemPager,
+) error {
+	if len(driveID) == 0 {
+		return clues.New("drive id is empty")
+	}
+
+	if refreshInterval < 0 {
+		return clues.New("invalid refresh interval")
+	}
+
+	if svc == nil {
+		return clues.New("nil graph servicer")
+	}
+
+	if itemPagerFunc == nil {
+		return clues.New("nil item pager")
+	}
+
+	return nil
+}
+
+// getItemProps returns the item properties for the specified drive item ID
+func (uc *urlCache) getItemProperties(
+	ctx context.Context,
+	itemID string,
+) (*itemProperties, error) {
+	if len(itemID) == 0 {
+		return nil, clues.New("item id is empty")
+	}
+
+	ctx = clues.Add(ctx, "drive_id", uc.driveID)
+
+	// Lazy refresh
+	if uc.needsRefresh() {
+		err := uc.refreshCache(ctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	url, err := uc.readCache(ctx, itemID)
+	if err != nil {
+		return nil, err
+	}
+
+	return url, nil
+}
+
+// needsRefresh returns true if the cache is empty or if refresh interval has
+// elapsed
+func (uc *urlCache) needsRefresh() bool {
+	uc.cacheLock.RLock()
+	defer uc.cacheLock.RUnlock()
+
+	return len(uc.urlMap) == 0 || time.Since(uc.lastRefreshTime) > uc.refreshInterval
+}
+
+// refreshCache refreshes the URL cache by performing a delta query.
+func (uc *urlCache) refreshCache(
+	ctx context.Context,
+) error {
+	// Acquire mutex to prevent multiple threads from refreshing the
+	// cache at the same time
+	uc.refreshMutex.Lock()
+	defer uc.refreshMutex.Unlock()
+
+	// If the cache was refreshed by another thread while we were waiting
+	// to acquire mutex, return
+	if !uc.needsRefresh() {
+		return nil
+	}
+
+	// Hold cache lock in write mode for the entire duration of the refresh.
+	// This is to prevent other threads from reading the cache while it is
+	// being updated page by page
+	uc.cacheLock.Lock()
+	defer uc.cacheLock.Unlock()
+
+	// Issue a delta query to graph
+	logger.Ctx(ctx).Debug("refreshing url cache")
+
+	err := uc.deltaQuery(ctx)
+	if err != nil {
+		return err
+	}
+
+	logger.Ctx(ctx).Debug("url cache refresh complete")
+	// Update last refresh time
+	uc.lastRefreshTime = time.Now()
+
+	return nil
+}
+
+// deltaQuery will perform a delta query on the drive. updateCache will be
+// called for each page of items returned by the delta query and the cache will
+// be updated with the results
+func (uc *urlCache) deltaQuery(
+	ctx context.Context,
+) error {
+	logger.Ctx(ctx).Debug("starting delta query")
+
+	_, _, _, err := collectItems(
+		ctx,
+		uc.itemPagerFunc(uc.svc, uc.driveID, ""),
+		uc.driveID,
+		"",
+		uc.updateCache,
+		map[string]string{},
+		"",
+		uc.errors,
+	)
+	if err != nil {
+		return clues.Wrap(err, "delta query failed").WithClues(ctx)
```

Comment:
WithClues not required, the called function should already attach it, since you pass in the ctx.
```suggestion
		return clues.Wrap(err, "delta query failed")
```

Reply:
Thanks!